Fix/php81

Donatello-za opened the pull request:
Hi, I am concerned about the future of this repository and through this pull request I'm hoping to get the library up to PHP 8.1's standard and compatibility.

Some notes:

* These changes are not backwards compatible with PHP versions older than 8.1.
* The changes made with this [un-merged pull request GH-43](https://github.com/vanilla/garden-cli/pull/43) was also implemented.
* You'll have to consider strategies for how to allow upgrades from the 2.x branch or the 3.x branch.
* Every file apart from the unit tests have been worked on. This was to fully implement strong typing and PHP 8.1 features.
* The deprecated `LogFormatter` class and unit tests have been removed.
* Unit Testing success is 100% and coverage is the same as before. However, the dev dependencies of the external `vanilla/garden-container` repository causes deprecation warnings during unit testing. To solve this, that package will also need upgrading to PHP 8.1
* Psalm analyses succeeds 100% with 0 warnings or notices.

If you end up using this pull request your welcome to squash the commits, I can also do it from this side but I'm not sure if you'd like to see the progression of the commits.

Donatello-za committed:
Initial PHP 8.1 compatibility fixes.

Donatello-za committed:
Fixed compatibility return type for OptSchema::jsonSerialize().

Donatello-za committed:
PHP 8.1 upgrades for Cli and Args classes.

Donatello-za committed:
PHP 8.1 upgrades for StreamLogger, Table and TaskLogger classes.

Donatello-za committed:
PHP 8.1 upgrades for CliApplication and Identifier classes.

Donatello-za committed:
PHP 8.1 upgrades for CommandSchema, MetaTrait and OptSchema classes.

Donatello-za committed:
PHP 8.1 Fixed various failed test cases.

Donatello-za committed:
PHP 8.1 Fixed various Psalm warnings.

Donatello-za committed:
PHP 8.1 Removed deprecated LogFormatter class and tests.

Donatello-za committed:
PHP 8.1 Updated README.md for upgrade tips.

Donatello-za committed:
PHP 8.1 Minor README.md formatting fixes.

Donatello-za commented:
I did notice just now that some source code formatting changes crept in due to my IDE setup. If you are not happy with it I can change those back.

thisispiers committed:
Update psr/log to ^2.0

blacktek commented:
Hello,
  any idea when the pull request of Donatello-za can be accepted and the new package released with composer?

Donatello-za committed:
Merge pull request #1 from thisispiers/fix/php81

Update psr/log to ^2.0

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/vanilla/garden-cli?pullRequest=44) <br/>All committers have signed the CLA.

blacktek commented on composer.json:
```diff
@@ -11,7 +11,7 @@
 	"require": {
```
I would suggest to use a require for "php 7.4" as the code is basically compatible with 7.4

Donatello-za replied:
Hi @blacktek, unfortunately it is not backwards compatible with PHP 7.4 due to keywords such `mixed` being used as parameters or return types amongst other things.

blacktek replied:
I'm currently using it on 7.4 too and is working. 7.4 doesn't support multiple union return types (e.g. string | bool), that I remember were not used. Are you sure that those are currently used?

Donatello-za replied:
I assumed there would be problems since the psalm test fails but if these warnings/errors can be fixed quickly in order to return composer.json to 7.4 it may be worth it. I will look into it when I have some time.

Donatello-za replied:
If you are struggling to set up my fork in `composer.json` here is an extract you can use:

```json
{
  "require": {
    "php": "^8.1",
    "vanilla/garden-cli": "dev-fix/php81"
  },
  "repositories": [
    { "type": "vcs", "url":  "https://github.com/donatello-za/garden-cli" }
  ]
}
```

blacktek replied:
tnx a lot :) this is exactly what I was using.

Donatello-za committed:
PHP 8.1 Composer update together with psalm fixes.

Donatello-za committed:
PHP 8.1 Fixed error when both global and command descriptions supplied.

The OptSchema::merge() function may create an array of strings if a "main"/"global" description is supplied as well as a description for individual commands. This fix will prevent errors in such a case and will simply show the correct description for the command.
